lib/convex/context.ex

```diff
@@ -0,0 +1,1942 @@
+    %Ctx{dir: dir, mod: mod, auth: auth, sess: sess, policy: policy} = ctx
+    %Ctx{state: :standby, dir: dir, mod: mod, sub: nil,
+         auth: auth, sess: sess, policy: policy}
+  end
+
+
+  @spec reset(context :: This.t) :: context :: This.t
+  @doc """
+  Resets a context keeping **ONLY** the callback module and director.
+
+  The returned context can only be used after recasting.
+  """
+
+  def reset(%Ctx{} = ctx) do
+    %Ctx{dir: dir, mod: mod} = ctx
+    %Ctx{state: :reset, dir: dir, mod: mod}
+  end
+
+
+  @spec merge(base_context :: This.t, context :: This.t) :: context :: This.t
+  @doc """
+  Merges authentication, session, policy, NEW assignment and tags
+  from a given context into base context. Contexts can be of different type
+  (different callback module).
+  """
+
+  def merge(%Ctx{} = base, %Ctx{} = from) do
+    %Ctx{base |
+      auth: from.auth,
+      sess: from.sess,
+      policy: from.policy,
+    }
+      |> assign_new(from.assigns)
+      |> add_tags(from.tags)
+  end
+
+
+  @spec perform(context :: This.t, pipeline :: Pipeline.t) :: context :: This.t
+  @spec perform(context :: This.t, pipeline :: Pipeline.t, options ::  Keyword.t)
+    :: context :: This.t
+  @doc """
+  Performs a pipeline without exception.
+
+  The pipeline could be created using `Convex.Pipeline.prepare/1`.
+
+  The options are passed to the callback module function `pipeline_performed/3`.
+  """
+
+  def perform(ctx, pipeline, opts \\ [])
+
+  def perform(%Ctx{state: :standby} = ctx, [{_, _, _, _} | _] = pipe, opts) do
+    _start_performing(ctx, pipe, opts)
+  end
+
+  def perform(%Ctx{state: :standby}, _pipe, _opts) do
+    raise ArgumentError, message: "Invalid pipeline"
+  end
+
+  def perform(%Ctx{}, _pipe, _opts) do
+    raise ArgumentError, message: "Context pipeline's is already running"
+  end
+
+
+  @spec perform!(context :: This.t, pipeline :: Pipeline.t) :: context :: This.t
+  @spec perform!(context :: This.t, pipeline :: Pipeline.t, options: Keyword.t)
+    :: context :: This.t
+  @doc """
+  Performs a pipeline raising an exception in case of failure and returning
+  any possible result directly.
+
+  The pipeline could be created using `Convex.Pipeline.prepare/1`.
+
+  The options are passed to the callback module function `pipeline_performed!/3`.
+  """
+
+  def perform!(ctx, pipeline, opts \\ [])
+
+  def perform!(%Ctx{state: :standby} = ctx, [{_, _, _, _} | _] = pipe, opts) do
+    _start_performing!(ctx, pipe, opts)
+  end
+
+  def perform!(%Ctx{state: :standby}, _pipe, _opts) do
+    raise ArgumentError, message: "Invalid pipeline"
+  end
+
+  def perform!(%Ctx{}, _pipe, _opts) do
+    raise ArgumentError, message: "Context pipeline's is already running"
+  end
+
+
+  @spec resume(context :: This.t, operation :: This.op, args :: map)
+    :: context :: This.t
+  @doc """
+  Resumes the execution of an operation. Useful if the operation was moved
+  around to another node and should be resumed there.
```

Comment:
How can an operation be moved to another node? Maybe an example of when this would be used could be useful.

Reply:
I'll try to be more specific... maybe this sounds too grand for what it is. I just mean that if you send the context to another node and want to bootstrap the current operation you can use this...